src/core/tsi/ssl_transport_security.cc

```diff
@@ -1538,15 +1539,42 @@ static tsi_result ssl_handshaker_next(tsi_handshaker* self,
   // If there are received bytes, process them first.
   tsi_ssl_handshaker* impl = reinterpret_cast<tsi_ssl_handshaker*>(self);
   tsi_result status = TSI_OK;
-  size_t bytes_consumed = received_bytes_size;
   size_t bytes_written = 0;
   if (received_bytes_size > 0) {
-    status = ssl_handshaker_process_bytes_from_peer(impl, received_bytes,
-                                                    &bytes_consumed, error);
-    while (status == TSI_DRAIN_BUFFER) {
-      status = ssl_handshaker_write_output_buffer(self, &bytes_written, error);
-      if (status != TSI_OK) return status;
-      status = ssl_handshaker_do_handshake(impl, error);
+    unsigned char* remaining_bytes_to_write_to_network_io =
+        const_cast<unsigned char*>(received_bytes);
+    std::size_t remaining_bytes_to_write_to_network_io_size =
+        received_bytes_size;
+    std::size_t number_bio_write_attempts = 0;
+    while (remaining_bytes_to_write_to_network_io_size > 0 &&
+           (status == TSI_OK || status == TSI_INCOMPLETE_DATA) &&
+           number_bio_write_attempts < TSI_SSL_MAX_BIO_WRITE_ATTEMPTS) {
+      number_bio_write_attempts++;
+
+      // Try to write all of the remaining bytes to the BIO.
+      std::size_t bytes_written_to_network_io =
+          remaining_bytes_to_write_to_network_io_size;
+      status = ssl_handshaker_process_bytes_from_peer(
+          impl, remaining_bytes_to_write_to_network_io,
+          &bytes_written_to_network_io, error);
+
+      // As long as the BIO is full, drive the SSL handshake to consume bytes
+      // from the BIO. If the SSL handshake returns any bytes, write them to the
+      // peer.
+      while (status == TSI_DRAIN_BUFFER) {
+        status =
+            ssl_handshaker_write_output_buffer(self, &bytes_written, error);
+        if (status != TSI_OK) return status;
+        status = ssl_handshaker_do_handshake(impl, error);
+      }
+
+      // Move the pointer to the first byte not yet successfully written to the
+      // BIO.
+      remaining_bytes_to_write_to_network_io_size -=
+          bytes_written_to_network_io;
+      if (remaining_bytes_to_write_to_network_io_size > 0) {
```

Comment:
If we do that, then on the last write we will push the `remaining_bytes_to_write_to_network_io` out of the allowed range `[received_bytes, received_bytes + received_bytes_size]`. If the pointer was used after that, then it would be undefined behavior.

Reply:
Yup fair point, updated.